### PR TITLE
FireSim Integration

### PR DESCRIPTION
**Pull request by davidbiancolin**
This is the first step towards making FireSim a better library. To use FireSim out of REBAR you'll still need to use targets in the firechip SUB_PROJECT for now. 

Overview: 
- Adds FireChip as a separate SUB_PROJECT; target-level RTL simulation is currently broken since i have not yet provided a test harness.
- Modifies the existing init scripts, and adds a firesim-setup.sh that wraps build-setup.sh in firesim. 
- Modifies build.sbt to include FireSim-provided deps, and adds the RC hack to allow us to use midas annotations in RC source 

From FireSim:
- Moves most of the code under src/main/scala/firesim into REBAR.
  - includes Targets.scala, TargetConfigs.scala, Generator.scala, and the scalatest suite

From FireChip:
- Moves some target-land tests for the NIC and blkdev into tests/

### CODE REVIEW

File: build.sbt [hunk 1]
```diff
@@ -9,28 +11,24 @@ lazy val commonSettings = Seq(
     case _ => MergeStrategy.first}},
   scalacOptions ++= Seq("-deprecation","-unchecked","-Xsource:2.11"),
   libraryDependencies += "org.scalatest" %% "scalatest" % "3.0.5" % "test",
-  libraryDependencies += "org.json4s" %% "json4s-native" % "3.6.1",
+  libraryDependencies += "org.json4s" %% "json4s-jackson" % "3.6.1",
   libraryDependencies += "org.scala-lang" % "scala-reflect" % scalaVersion.value,
-  libraryDependencies += "edu.berkeley.cs" %% "firrtl-interpreter" % "1.2-SNAPSHOT",
+  //libraryDependencies += "edu.berkeley.cs" %% "firrtl-interpreter" % "1.2-SNAPSHOT",
```
Comment (abejgonzalez) [inline build.sbt]: Just remove no?

Comment (davidbiancolin) [reply build.sbt]: Yeah. Do we use it anywhere?

Comment (abejgonzalez) [reply build.sbt]: I don't and nor does the BOOM team? Others maybe? @colinschmidt @zhemao @azidar

Comment (colinschmidt) [reply build.sbt]: Uh at some point barstools used this

Comment (davidbiancolin) [reply build.sbt]: Alright, should be easy to reintroduce.

Comment (abejgonzalez) [reply build.sbt]: Will it be used again? @colinschmidt @jwright6323

File: generators/firechip/src/main/scala/Generator.scala [hunk 2]
```diff
@@ -0,0 +1,133 @@
+//See LICENSE for license details.
```
Comment (abejgonzalez) [inline generators/firechip/src/main/scala/Generator.scala]: This file seems to be executable. `-rwxrwxr-x`. Can you confirm if this is the case on your end. And if so, can we switch it to something like `-rw-rw-r--`

File: generators/firechip/src/main/scala/Targets.scala [hunk 3]
```diff
@@ -0,0 +1,176 @@
+package firesim.firesim
```
Comment (abejgonzalez) [inline generators/firechip/src/main/scala/Targets.scala]: This file also seems to be executable. `-rwxrwxr-x`. Can you confirm if this is the case on your end? And if so, can we switch it to something like `-rw-rw-r--`

File: scripts/init-submodules-no-riscv-tools.sh [hunk 4]
```diff
@@ -6,14 +6,27 @@ set -o pipefail
...
 unamestr=$(uname)
 RDIR=$(pwd)
+scripts_dir="$( cd "$( dirname "${BASH_SOURCE[0]}" )" >/dev/null 2>&1 && pwd )"
 
 # ignore riscv-tools for submodule init recursive
 # you must do this globally (otherwise riscv-tools deep
 # in the submodule tree will get pulled anyway
 git config --global submodule.riscv-tools.update none
 git config --global submodule.esp-tools.update none
 git config --global submodule.experimental-blocks.update none
```
Comment (abejgonzalez) [inline scripts/init-submodules-no-riscv-tools.sh]: So just to make sure I understand correctly, the user must init the rebar submodules first, then they run the firesim init script. How do we want to document this (put this in the rebar docs? and firesim docs) or does FireSim handle submodule init differently?

**Comment by abejgonzalez** [comment]
LGTM. Runs SmallBoomConfig nicely! Testing on Default for now.

**Comment by abejgonzalez** [comment]
Just waiting on comments being addressed then I think this can be merged in.

File: build.sbt (hunk 1 above)
Comment (colinschmidt) [reply build.sbt]: Its currently used to run tests for macro compiler: https://github.com/ucb-bar/barstools/blob/master/build.sbt#L34

Comment (abejgonzalez) [reply build.sbt]: Do the tests run without this right now if you pull this?

File: scripts/firesim-setup.sh [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+RDIR=$(pwd)
+scripts_dir="$( cd "$( dirname "${BASH_SOURCE[0]}" )" >/dev/null 2>&1 && pwd )"
+sims_dir=$scripts_dir/../sims/
+
+# Reenable the FireSim submodule
+git config --unset submodule.sims/firesim.update || true
+cd $sims_dir
+git submodule update --init firesim
+cd firesim
+./build-setup.sh $@ --library
```
Comment (jerryz123) [inline scripts/firesim-setup.sh]: This seems to rebuild riscv-tools. Is this necessary, given that REBAR already packages all the tools?

Comment (davidbiancolin) [reply scripts/firesim-setup.sh]: I figure that most people will use `./firesim-setup.sh --fast`, to get a prebuilt toolchain on AWS. I have yet to add a no-toolchain feature to build-setup.sh, but should do that in the future.

File: scripts/init-submodules-no-riscv-tools.sh (hunk 4 above)
Comment (davidbiancolin) [reply scripts/init-submodules-no-riscv-tools.sh]: Yeah, that's correct. I figure that in many cases people won't want to use firesim (they might be doing local dev, for instance). 

REBAR docs; build-setup.sh has a different flow when initializing firesim-as-top.

File: generators/firechip/src/main/scala/Targets.scala (hunk 3 above)
Comment (davidbiancolin) [reply generators/firechip/src/main/scala/Targets.scala]: I will fix this when i get back to my machine.

File: generators/firechip/src/main/scala/Generator.scala (hunk 2 above)
Comment (davidbiancolin) [reply generators/firechip/src/main/scala/Generator.scala]: As below.

File: build.sbt (hunk 1 above)
Comment (davidbiancolin) [reply build.sbt]: I can just fix this when i get back to my machine.

File: variables.mk [hunk 6]
```diff
@@ -74,6 +74,18 @@ ifeq ($(SUB_PROJECT),hwacha)
 	TB                ?= TestDriver
 	TOP               ?= ExampleRocketSystem
 endif
+# Stand-in firechip variables
```
Comment (abejgonzalez) [inline variables.mk]: @zhemao What should these be to simulate `firechip` in Chipyard?

Comment (davidbiancolin) [reply variables.mk]: We've resolved to comment this out temporarily.

**Pull request merged**